index.d.ts

```diff
@@ -0,0 +1,18 @@
+import {Middleware, Dispatch} from "redux";
+
+
+export type ThunkAction<R, S, E> = (dispatch: Dispatch<S>, getState: () => S,
+                                    extraArgument: E) => R;
+
+declare module "redux" {
+  export interface Dispatch<S> {
+    <R, E>(asyncAction: ThunkAction<R, S, E>): R;
```

Comment:
I see. That make sense.

I'm experimenting alternative approach that `Store` is not typed with `<S>`. Only `reduce` and `getState` is typed with `<S>` because the producer (which `reduce`) and consumer (which `getState`) have separate concerns.